frame/contracts/src/wasm/runtime.rs

```diff
@@ -594,6 +602,35 @@ where
 		Ok(())
 	}
 
+	/// Recovers ECDSA compressed public key from the signature and message hash.
+	#[cfg(feature = "unstable-interface")]
+	fn ecdsa_recover(
+		&mut self,
+		signature_ptr: u32,
+		message_hash_ptr: u32,
+		output_ptr: u32,
+	) -> Result<ReturnCode, DispatchError> {
+		self.charge_gas(RuntimeCosts::EcdsaRecovery)?;
+
+		let mut signature: [u8; 65] = [0; 65];
+		self.read_sandbox_memory_into_buf(signature_ptr, &mut signature)?;
+		let mut message_hash: [u8; 32] = [0; 32];
+		self.read_sandbox_memory_into_buf(message_hash_ptr, &mut message_hash)?;
+
+		let result = self.ext.ecdsa_recover(&signature, &message_hash);
+
+		match result {
+			Ok(pub_key) => {
+				// Write the recovered compressed ecdsa public key back into the sandboxed output
+				// buffer.
+				self.write_sandbox_memory(output_ptr, pub_key.as_ref())?;
+
+				Ok(ReturnCode::Success)
+			},
+			Err(_) => Ok(ReturnCode::RecoveryFailed),
+		}
```

Comment:
Honest question: Do you think having a return code rather than trapping on error is useful here? I am asking because there is really no proper guideline when to trap and when to allow the contract to handle the error. Most functions do trap on error, though. @cmichi What are your thoughts?

Reply:
Initially, I implemented it without a return code and throw `ContractTrapped`. But during testing in ink! contract found, that it can be useful during debugging to understand that `ecdsa_recover` returned an error.

Reply:
If we decided to let it trap instead of returning an error we wouldn't throw `ContractTrapped` but create a new error so that it is obvious what happened. The only question is if it is useful for the calling contract to recover from the error. Debugging should not play a role in this decisions.

Reply:
>I am asking because there is really no proper guideline when to trap and when to allow the contract to handle the error. 

Yeah all of those trap vs. return err decisions are definitely a grey area. 

>If we decided to let it trap instead of returning an error we wouldn't throw `ContractTrapped` but create a new error so that it is obvious what happened.

So you would create a new error `EcdsaRecoveryFailed`? 

I like the idea of returning the error back to the developer better, since it keeps the business logic of the contract in the contract. It should not be relevant to the outside world if this particular `contracts` API function failed and what it failed with, the developer should be able to decide on its own what error he wants to expose as public in this case.

That's no strong opinion though and more of a gut feeling. Since we are not sure about how developers will actually use this particular function we might just make it consistent with [the Ethereum world](https://docs.openzeppelin.com/contracts/2.x/api/cryptography#ECDSA-recover-bytes32-bytes-):

> This call does not revert if the signature is invalid, or if the signer is otherwise unable to be retrieved. In those scenarios, the zero address is returned.

This also keeps more options open for developers, if we trap immediately it takes away any possibility of recovering. So since we're not sure what developers will use it for, I think it's better to allow the option for recovery.

Reply:
> So you would create a new error EcdsaRecoveryFailed?

In case we'd go the trap/revert route for this API (as opposed the the currently implemented return code).

I agree with the rest. We should keep it as-is then.